feat: ui primitive api autodocs

thatsamsonkid opened the pull request:
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our
      guidelines: https://github.com/goetzrobin/spartan/blob/main/CONTRIBUTING.md#-commit-message-guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [x] Other... Please describe:

## Which package are you modifying?

- [ ] accordion
- [ ] alert
- [ ] alert-dialog
- [ ] aspect-ratio
- [ ] avatar
- [ ] badge
- [ ] button
- [ ] calendar
- [ ] card
- [ ] checkbox
- [ ] collapsible
- [ ] combobox
- [ ] command
- [ ] context-menu
- [ ] data-table
- [ ] date-picker
- [ ] dialog
- [ ] dropdown-menu
- [ ] hover-card
- [ ] icon
- [ ] input
- [ ] label
- [ ] menubar
- [ ] navigation-menu
- [ ] pagination
- [ ] popover
- [ ] progress
- [ ] radio-group
- [ ] scroll-area
- [ ] select
- [ ] separator
- [ ] sheet
- [ ] skeleton
- [ ] slider
- [ ] sonner
- [ ] spinner
- [ ] switch
- [ ] table
- [ ] tabs
- [ ] textarea
- [ ] toast
- [ ] toggle
- [ ] tooltip
- [ ] typography

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Closes #247 

## What is the new behavior?

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information

thatsamsonkid commented:
Wanted you all's opinion on this. So I think I got a semi-decent executor to autodoc the UI primitives API. 

Output is a JSON file with an object where each property is the class name for a given component/directive with the following structure

```
"BrnAccordionItemDirective": {
    "file": "libs/ui/accordion/brain/src/lib/brn-accordion.directive.ts",
    "inputs": [
      {
        "name": "isOpened",
        "type": "boolean",
        "description": "Allows to set initial state of accordion"
      }
    ],
    "outputs": []
  },
```
it would require us to start self document each input output like so (it's currently not doing so well without explicit type on signal and assuming any in that case)

```
/**
* Allows to set initial state of accordion
*/
public readonly isOpened = input<boolean>(false);
```

my hope is we won't have to maintain the UI api docs as it's own task and we can just make sure it's well doc'ed in the code. Curious to hear your thoughts. My plan was place the static JSON in the site app's public assets folder and just read it in based on the className to feed them into a table on the page

@ashley-hunter @elite-benni  @goetzrobin

ashley-hunter commented:
Personally, I have been using this approach in projects for many years and it has worked well. The benefit of this approach is users also get docs in their editor, for example if I highlight over a property or ctrl click to definition I can see docs right inline in my editor as well as them being available in a docs site. And it also means the docs are derived from the code so less likely to be out of date. So I'm all for this approach!

I haven't looked into your code much, and it may do everything we need, but if not, just so you're aware, there are some solutions out there that are available if you don't want to write your own from scratch. One is compodoc (https://compodoc.app/) which actually comes built in with the Nx storybook executor, you can just enable it and get it to emit JSON. It's not perfect, but not bad. 

Another option which could be used within our own executor like you have been doing but may help extract information is Angular's own API extraction tooling - although it's not really documented but it's pretty good:

https://github.com/angular/angular/tree/main/packages/compiler-cli/src/ngtsc/docs/src

I really like the idea, just wanted to share these in case they are of any use to you!

thatsamsonkid commented:
> Personally, I have been using this approach in projects for many years and it has worked well. The benefit of this approach is users also get docs in their editor, for example if I highlight over a property or ctrl click to definition I can see docs right inline in my editor as well as them being available in a docs site. And it also means the docs are derived from the code so less likely to be out of date. So I'm all for this approach!
> 
> I haven't looked into your code much yet (as I'm on my phone), and it may do everything we need, but if not, just so you're aware, there are some solutions out there that are available if you don't want to write your own from scratch. One is compodoc (https://compodoc.app/) which actually comes built in with the Nx storybook executor, you can just enable it and get it to emit JSON. It's not perfect, but not bad.
> 
> Another option which could be used within our own executor like you have been doing but may help extract information is Angular's own API extraction tooling - although it's not really documented but it's pretty good:
> 
> https://github.com/angular/angular/tree/main/packages/compiler-cli/src/ngtsc/docs/src
> 
> I really like the idea, just wanted to share these in case they are of any use to you!

Yea I was thinking to myself afterwards I was like there's probably already solutions for this 😅, but i'm all for either. We can leverage compodoc if we prefer or I can work to probably make this more robust by leveraging the angular one you mentioned. I wasn't aware of it so thanks for mentioning them!

ashley-hunter commented:
> Yea I was thinking to myself afterwards I was like there's probably already solutions for this 😅, but i'm all for either. We can leverage compodoc if we prefer or I can work to probably make this more robust by leveraging the angular one you mentioned. I wasn't aware of it so thanks for mentioning them!

You may not need either if you have it working fine already 😊 it may be easier not to rely on an additional tool at all.

I guess we could also extract some additional things like the selector and exportAs values, which could also be useful in the docs.

elite-benni commented:
I am a big fan of keeping things together. 
If we can generate documentation from the comment, that is also used by the ide this would be very nice. 
If you always need to think about updating the api in the documentation page, it has a high probability to get out of synch.

I have nothing to add to Ashleys comments, because I have pretty much the same opinion.

thatsamsonkid commented:
Awesome sounds good, I guess since i'm this far let me see if I can also add grabbing the selector and exportAs if available, great suggestion @ashley-hunter. Thanks all for the input, just wanted to make sure we were good with starting to self document this in the code. I know the trade off is slightly "busier" looking code with the extra comments but I think it's better than risking having out of date api docs

thatsamsonkid commented:
Updated the structure to be like below so we can easily pull all the "accordion" api's for example and easily loop and create brain and helm sections on the page docs and dynamically create tables for each respective component/directive
```
{
  "accordion": {
    "brain": {
      "BrnAccordionContentComponent": {
        "file": "libs/ui/accordion/brain/src/lib/brn-accordion-content.component.ts",
        "inputs": [],
        "outputs": [],
        "selector": "brn-accordion-content",
        "exportAs": null
      },
   },
   "helm":{
      "HlmAccordionContentComponent": {
        "file": "libs/ui/accordion/helm/src/lib/hlm-accordion-content.component.ts",
        "inputs": [
          {
            "name": "userClass",
            "type": "ClassValue",
            "description": ""
          }
        ],
        "outputs": [],
        "selector": "hlm-accordion-content",
        "exportAs": null
      },
},
...
```

thatsamsonkid commented:
Just wanted to give an update I think I mainly have the few things below left for this:
- [ ] Fix some typing issues
- [ ] Add Tests
- [ ] Ensuring the docs get generated on app build
- [ ] Fixing an issue with the page nav stickyness
- [ ] Updating all primitive pages with the api docs sections and extra data parameters

For this last one should I just go ahead and add the sections to all the primitive pages as part of this and fix any issues we find or should we go one by one after merging this to make sure the docs for each primitive are complete and correct first

thatsamsonkid commented:
I'll resolve the check failures as soon as i can. 

But just to give an idea of whats happening.

- Created an executor that loops through all the brain and ui components and directives and uses ts-morph library to extract the metadata
- Exports all the metadata into a JSON and outputs it in public/data in app
- Loads this static json serverside on components layout page
- Each primitive page added brain and helm headings where appropriate so page-nav component picks up the headings
- adds an api table under each heading and leverages a new "api" metadata tag to determine the name of ui api data we should use
- if page has this "api" metadata parameter then page nav loads the additional subheadings mapping to all the components/directives for the primitive.

Let me know if there's any improvements anyone might have for this

thatsamsonkid commented:
There is like 1 or 2 components where this doesn't quite work as well, example being combox where because it itself doesn't have it's own components/directives but instead relies on a mix of our other primitives so it wouldn't have the api listed there but I figured we can add links to the related pages. 

For menubar and context menu it's a little duplicative (maybe doesn't matter since pulling from same docs source) but both pages have the menu api in this case. 

So just thought i'd point these 2 scenarios out. Other than that I think this may be good, let me know if anything otherwise

ashley-hunter commented:
Hey! Great work, works really well, such an improvement to the docs. I'll point out some of the areas that I encountered that I guess we could improve upon - but they may be future enhancements, or just updates needed in the components themselves to help extraction - but just raising them so we can discuss 😊

1. It doesn't seem to be able to infer the type of an input if not explicitly specified - is there anything we can do to improve this? E.g. check the default value for common types like string, boolean, number etc.. or I guess we could create an eslint rule to require a type? 

<img width="818" alt="image" src="https://github.com/user-attachments/assets/09a3ef21-bf79-4c6e-b66e-8c4ad66f0967" />

2. It would be good if we could detect the alias and show it instead as that would be what is used in the template by the consumer:

<img width="783" alt="image" src="https://github.com/user-attachments/assets/e162db40-922f-4433-96a3-86869aabd287" />

3. Type information - I suspect this would be a future enhancement and something that probably wouldn't be easy, but it would be good if we can provide more informative information about the values of some of the types:

<img width="801" alt="image" src="https://github.com/user-attachments/assets/372acca6-b023-4cfe-95d2-e646c15231c3" />

4. It seems to include casting in the value if used:

<img width="816" alt="image" src="https://github.com/user-attachments/assets/fb354129-8824-4506-8d92-5c558e95a909" />

5. I wonder is there any way to better handle column alignment here - do we truncate long values and show a tooltip?

<img width="817" alt="image" src="https://github.com/user-attachments/assets/025e9745-8287-4bad-99c2-8719553a212e" />

6. It doesn't seem to detect `model` as inputs or outputs, for example, calendar should have a `date` input and `dateChange` output:

<img width="903" alt="image" src="https://github.com/user-attachments/assets/664827bf-bd6f-491e-b197-de10ac8eb067" />

Regardless, we can pick any of these up at a later point if needed - this is still a great improvement to the docs! Nice job!

thatsamsonkid commented:
This is great feedback, thanks for taking a thorough look at it @ashley-hunter I definitely needed some fresh eyes to catch some of these

1. I think I can check if we can do better at inferring but if not yea we might need to just be more dilligent about typing it for doc purposes, an eslint rule could definitely be helpful. In general i think we might need to look at each primitive and refine the docs for them. I was consideing doing a issue for each but i didnt want to blow our milestone progress out of proportion.

2. good catch! Lemme see what i can do

3. Do you have an example of what you would want for this? I didnt quite get this one.

4. Let check this also

5. yea the column management might be a bit difficult if we dont do like a tooltip situation but I think i recall @goetzrobin was anti haha. Anyone have any other suggestions, otherwise we may have to go with the tooltips.

6. Another good catch I totally overlooked that

ashley-hunter commented:
Apologies, my description for No. 3 was not great - this is probably going to be the most difficult, and would maybe require the type checker, so not a big deal if it's out of scope, but for things like BadgeVariants['variant'] it would be useful if we could display the actual variants, e.g. 'default' | ' secondary' | 'destructive' | 'outline' for example.

thatsamsonkid commented:
Ahh gotcha thanks that makes sense and yea definitely would be a lot more useful. We could provide it as part of description maybe? Although kinda sucks this wouldn't be inferred and instead determined by our comments

ashley-hunter commented:
I guess we could consider specifying it as part of the JSDoc? That way it could be extracted and not actually part of the description column, e.g.:

```ts
	/**
	 * The variant of the badge.
	 * @type {'default' | 'secondary' | 'destructive' | 'outline'}
	 * @default 'default'
	 */
	public readonly variant = input<BadgeVariants['variant']>('default');
```

thatsamsonkid commented:
This looks pretty promising actually. I think in most cases we probably only have a handful of values so maintaining them in JSDoc might not be too bad. I can take this approach

goetzrobin commented:
Also definitely didn't want to be any tooltip 😩 seems like it would be good to hide some of the more detailed info?

elite-benni commented on apps/app/src/app/shared/layout/ui-docs-section/ui-docs-section.component.ts:
```diff
@@ -0,0 +1,92 @@
+		@if (componentDocs() && componentEntries() && componentEntries().length > 0) {
+			@for (entry of componentEntries(); track entry) {
+				<h4 class="${hlmH4} mb-2 mt-6 pt-12" [id]="entry.toLowerCase()">{{ entry }}</h4>
+				<p>Selector: {{ componentItems()[entry].selector }}</p>
```
I was thinking about packing the selector into a code section.
This way the selector gets a little bit more focus.

```<code class="${hlmCode}">{{ componentItems()[entry].selector }}</code>```

thatsamsonkid replied:
<img width="706" alt="Screenshot 2025-05-10 at 10 51 31 PM" src="https://github.com/user-attachments/assets/b60d5db3-9a10-4640-85ed-a8856961601e" />
Updated, let me know if any thing else

elite-benni reviewed (approved):
Wow!
Pretty amazing work, it really improves the documentation a lot. 
Just added a few small comments, that are definitely no blockers.
Ashley already did an great feedback with many examples! ;)

thatsamsonkid commented on apps/app/src/app/pages/(components)/components/(accordion)/accordion--multiple-opened.example.ts:
```diff
@@ -26,7 +26,7 @@ import { HlmIconDirective } from '@spartan-ng/ui-icon-helm';
 	template: `
 		<div hlmAccordion type="multiple" class="pb-4">
-			<div hlmAccordionItem isOpened>
+			<div hlmAccordionItem [isOpened]="true">
```
I think I was getting some warning or something I don't quite remember

thatsamsonkid commented on apps/app/src/app/pages/(components)/components/(context-menu)/context-menu.page.ts:
```diff
@@ -68,6 +70,12 @@ export const routeMeta: RouteMeta = {
 				<spartan-code [code]="defaultSkeleton" />
 			</div>
 
+			<spartan-section-sub-heading id="brn-api">Brain API</spartan-section-sub-heading>
```
Yea i started with that approach but the problem I ran into is the current implementation for the page-nav is expecting the headings to be explicitely available on the respective page template. So I opted for this instead it was easier to already have the headings there and append the subheadings to it. Maybe I missed an alternative solution to that but I didn't want to change the way the page-nav was working today

goetzrobin commented:
@thatsamsonkid do you need anything from me to get this over the finish line? Looks like this will be a really really good addition! Looking forward to it! Also, @ashley-hunter  @thatsamsonkid @elite-benni feel free to merge whenever you guys think its good to go!

thatsamsonkid commented:
@goetzrobin @ashley-hunter @elite-benni - I added the selector display update, added support for models and updated the script to prefer the alias name of inputs if applicable. I also did some quick fixes to the table display for mobile breakpoint and added some fixed column widths to mitigate the column misalignment for the most part and should be less severe. For the other items Ashley mentioned I think I definitely will need some more time to implement and we can create some issues around them. This PR is already pretty big and rebasing is becoming a pain 😭 Feel free to give this another quick look if everyone's good i'd say lets merge!

elite-benni commented:
Nice, it really looks good and clean now!


A small cosmetic bug is that the value of the description sometimes breaks through the column.
![image](https://github.com/user-attachments/assets/4ca8f49d-c391-4329-9c8b-b25d3b71a9c4)
Maybe we should ``truncate`` or ``break-all`` the values. 
I would vote for break-all, it does not happen that often, and with break we won't hide the information.


Some points for future improvements imho would be: 

- Somehow parse and list the Variants (already mentioned)
- Some components use the provideConfig Pattern, maybe it would be nice to have the Configurations documented.
- Genereal documentation to Directives and Components. 
(sometimes only the Directive with Selector is listed, which does not give much information)
- A lot of documentations are missing.
 

I think this PR can be merged, as it provides a lot of value and my mentioned points are only further improvements.

ashley-hunter commented:
Apologies it took me a few days to get back to this. I have nothing further to add, looks great! Excellent improvement, we can make additional incremental improvements going forward.

thatsamsonkid commented:
Awesome thanks all appreciate the reviews, I'll make issues for the remaining items mentioned. Merging!
